packages/nx/src/command-line/migrate.ts

```diff
@@ -158,146 +177,137 @@ export class Migrator {
-          this._updatePackageJson(u, packages[u], {
+        .map((packageName) =>
+          this._updatePackageJson(packageName, packages[packageName], {
             ...collectedVersions,
             [targetPackage]: target,
           })
         )
     );
-    return childCalls.reduce(
-      (m, c) => {
-        Object.keys(c).forEach((r) => {
-          if (!m[r] || this.gt(c[r].version, m[r].version)) {
-            m[r] = c[r];
+
+    return childPackageMigrations.reduce(
+      (migrations, childMigrations) => {
+        for (const migrationName of Object.keys(childMigrations)) {
+          if (
+            !migrations[migrationName] ||
+            this.gt(
+              childMigrations[migrationName].version,
+              migrations[migrationName].version
+            )
+          ) {
+            migrations[migrationName] = childMigrations[migrationName];
           }
-        });
-        return m;
+        }
+        return migrations;
       },
       {
         [targetPackage]: {
           version: migrationsJson.version,
           addToPackageJson: target.addToPackageJson || false,
         },
-      }
+      } as Record<string, PackageJsonUpdateForPackage>
     );
   }
 
   private collapsePackages(
     packageName: string,
     targetVersion: string,
-    m: MigrationsJson | null
-  ) {
+    migration: ResolvedMigrationConfiguration
+  ): Record<string, PackageJsonUpdateForPackage> {
     // this should be used to know what version to include
     // we should use from everywhere we use versions
 
-    if (packageName === '@nrwl/workspace') {
-      if (!m.packageJsonUpdates) m.packageJsonUpdates = {};
-      m.packageJsonUpdates[`${targetVersion}-defaultPackages`] = {
+    if (migration.packageGroup) {
+      migration.packageJsonUpdates ??= {};
+      migration.packageJsonUpdates[`${targetVersion}-defaultPackages`] = {
         version: targetVersion,
-        packages: [
-          'nx',
-          '@nrwl/angular',
-          '@nrwl/cypress',
-          '@nrwl/devkit',
-          '@nrwl/eslint-plugin-nx',
-          '@nrwl/express',
-          '@nrwl/jest',
-          '@nrwl/js',
-          '@nrwl/cli',
-          '@nrwl/linter',
-          '@nrwl/nest',
-          '@nrwl/next',
-          '@nrwl/node',
-          '@nrwl/nx-cloud',
-          '@nrwl/nx-plugin',
-          '@nrwl/react',
-          '@nrwl/storybook',
-          '@nrwl/web',
-          '@nrwl/react-native',
-          '@nrwl/detox',
-        ].reduce(
-          (m, c) => ({
-            ...m,
-            [c]: {
-              version: c === '@nrwl/nx-cloud' ? 'latest' : targetVersion,
+        packages: migration.packageGroup.reduce((acc, packageConfig) => {
+          const { package: pkg, version } =
+            typeof packageConfig === 'string'
+              ? { package: packageConfig, version: targetVersion }
+              : packageConfig;
+
+          return {
+            ...acc,
+            [pkg]: {
+              version,
```

Comment:
Talked to vsavkin a bit more, and there should probably be a recursive portion of this.

Collapse packages would need to support packages in the packageGroup, which also have a packageGroup themselves. There should also be some circular prevention. I.e.: 
```
a: ['b', 'c',]
b: ['a', 'd']
c: []
d: []
```

`nx migrate a` should expand into `a, b, c`, then b should be expanded into `b, a, d`, then `c` and `d` shouldn't expand since they have no group. The function would need to notice that its already expanded `a`, and refrain from doing so when it handles expanding `b`.

Does that make sense?

Reply:
Yeah, it does make sense, I'll make it recursive, and will try to make it a little more readable in the process, the code is pretty hard to read currently.

Reply:
@AgentEnder @vsavkin I added a UT before starting to work on this, and I was surprised to see the test pass 😂

I looked at the code and than I saw that `_updatePackageJson` already recursively calls itself, so this is already handled 👍🏻

I'll upload the test just so this will be covered.